Ignore 'extra' nodes when querying for field names of children by index

ChewyGumball opened the pull request:
`ts_node_field_name_for_child` returns incorrect field names when there are 'extra' nodes preceding the child at the requested index. The index stored in the field map is based on the grammar for the rule, but 'extra' nodes can appear anywhere, and increase the `child_index` of subsequent nodes.

Example:
```cpp
for(auto i : c) // help
i++;
```

labels the `//help` with the `body` field, not the `i++` expression statement when queried using `ts_node_field_name_for_child`:

(format is `[start line, start column] -> [end line, end column] {node name} ({field name})` with indentation before the node name representing nested nodes)
```
for(auto i : c) // help
i++;
[ 0,  0] -> [ 1,  4]     for_range_loop
[ 0,  0] -> [ 0,  3]         for
[ 0,  3] -> [ 0,  4]         (
[ 0,  4] -> [ 0,  8]         placeholder_type_specifier (type)
[ 0,  4] -> [ 0,  8]             auto
[ 0,  9] -> [ 0, 10]         identifier (declarator)
[ 0, 11] -> [ 0, 12]         :
[ 0, 13] -> [ 0, 14]         identifier (right)
[ 0, 14] -> [ 0, 15]         )
[ 0, 16] -> [ 0, 23]         comment (body)
[ 1,  0] -> [ 1,  4]         expression_statement
[ 1,  0] -> [ 1,  3]             update_expression
[ 1,  0] -> [ 1,  1]                 identifier (argument)
[ 1,  1] -> [ 1,  3]                 ++ (operator)
[ 1,  3] -> [ 1,  4]             ;
```

My attempt at a fix is to count the number of 'extra' nodes that come before the child at the requested index, then subtract that number from the index when comparing to the index in the field map. This is similar to how the cursor based version works. Now the above example shows:
```
for(auto i : c) // help
i++;
[ 0,  0] -> [ 1,  4]     for_range_loop
[ 0,  0] -> [ 0,  3]         for
[ 0,  3] -> [ 0,  4]         (
[ 0,  4] -> [ 0,  8]         placeholder_type_specifier (type)
[ 0,  4] -> [ 0,  8]             auto
[ 0,  9] -> [ 0, 10]         identifier (declarator)
[ 0, 11] -> [ 0, 12]         :
[ 0, 13] -> [ 0, 14]         identifier (right)
[ 0, 14] -> [ 0, 15]         )
[ 0, 16] -> [ 0, 23]         comment
[ 1,  0] -> [ 1,  4]         expression_statement (body)
[ 1,  0] -> [ 1,  3]             update_expression
[ 1,  0] -> [ 1,  1]                 identifier (argument)
[ 1,  1] -> [ 1,  3]                 ++ (operator)
[ 1,  3] -> [ 1,  4]             ;
```

https://github.com/tree-sitter/tree-sitter/issues/1642 has similar symptoms, however their examples don't include 'extra' nodes. I get the same incorrect labels on their example, so this PR does not close that issue:
```
struct Foo { static const int bar = 10; };
[ 0,  0] -> [ 0, 41]     struct_specifier
[ 0,  0] -> [ 0,  6]         struct
[ 0,  7] -> [ 0, 10]         type_identifier (name)
[ 0, 11] -> [ 0, 41]         field_declaration_list (body)
[ 0, 11] -> [ 0, 12]             {
[ 0, 13] -> [ 0, 39]             field_declaration
[ 0, 13] -> [ 0, 19]                 storage_class_specifier (type)
[ 0, 13] -> [ 0, 19]                     static
[ 0, 20] -> [ 0, 25]                 type_qualifier (declarator)
[ 0, 20] -> [ 0, 25]                     const
[ 0, 26] -> [ 0, 29]                 primitive_type
[ 0, 30] -> [ 0, 33]                 field_identifier (default_value)
[ 0, 34] -> [ 0, 35]                 =
[ 0, 36] -> [ 0, 38]                 number_literal
[ 0, 38] -> [ 0, 39]                 ;
[ 0, 40] -> [ 0, 41]             }
[ 0, 41] -> [ 0, 42]     ;
```

I am not sure how to write a test for this as there are no tests that check field names. In my own project that uses tree-sitter, this modification means I get correct field names in the presence of 'extra' nodes.

ChewyGumball commented:
Is there anything I need to do to help move this PR along?

ahlinc commented:
@ChewyGumball if you can [write a test](https://github.com/tree-sitter/tree-sitter/blob/master/cli/src/tests/node_test.rs) for your code it would be awesome, thanks for the ping.
@maxbrunsfeld could you review this PR?

ChewyGumball commented:
@ahlinc I didn't realize all these tests existed, so thanks for pointing out where this kind of test should be! I have added a new test that fails before my change, and succeeds afterwards. I also added a test for the case in https://github.com/tree-sitter/tree-sitter/issues/1642, though it still fails so I commented it out until it is fixed.
